From 08aff0dce57252bae09f501a3b2f0b25832c48e5 Mon Sep 17 00:00:00 2001
From: "Christopher R. Palmer" <crpalmer@gmail.com>
Date: Thu, 16 Mar 2017 21:33:27 -0400
Subject: [PATCH] linker: Allow devices to force shim libs

There are certain contexts in which the environment is cleansed.
Two examples that I know of are gps processes and mali gles blobs.

Generally, it is a better idea to use the environment variable because
then you can customize it on a per-service / script level and it is
easier to test and debug changes on the fly.

However, to avoid having libdimytry hexedit these blobs, allow a
last resort device level forced list of shim libraries.

Change-Id: I2f6aff9325beb5aa2f748bf72e6c3c0535d5aac2
---
 Android.mk        |  4 ++++
 linker/linker.cpp | 10 +++++++++-
 2 files changed, 13 insertions(+), 1 deletion(-)

diff --git a/Android.mk b/Android.mk
index 888404cf7..60a5db1fe 100644
--- a/Android.mk
+++ b/Android.mk
@@ -1,4 +1,8 @@
 LOCAL_PATH := $(call my-dir)
 
+ifneq ($(LINKER_FORCED_SHIM_LIBS),)
+    LOCAL_CFLAGS += -DFORCED_SHIM_LIBS="\"$(LINKER_FORCED_SHIM_LIBS)\""
+endif
+
 include $(call all-makefiles-under,$(LOCAL_PATH))
 
diff --git a/linker/linker.cpp b/linker/linker.cpp
index 104f22325..cd114f888 100644
--- a/linker/linker.cpp
+++ b/linker/linker.cpp
@@ -716,7 +716,7 @@ static void reset_g_active_shim_libs(void) {
   }
 }
 
-void parse_LD_SHIM_LIBS(const char* path) {
+void parse_shim_libs(const char* path) {
   g_ld_all_shim_libs.clear();
   if (path != nullptr) {
     // We have historically supported ':' as well as ' ' in LD_SHIM_LIBS.
@@ -731,6 +731,14 @@ void parse_LD_SHIM_LIBS(const char* path) {
   reset_g_active_shim_libs();
 }
 
+void parse_LD_SHIM_LIBS(const char* path) {
+  g_ld_all_shim_libs.clear();
+#ifdef FORCED_SHIM_LIBS
+  parse_shim_libs(FORCED_SHIM_LIBS);
+#endif
+  parse_shim_libs(path);
+}
+
 template<typename F>
 static void for_each_matching_shim(const char *const path, F action) {
   if (path == nullptr) return;
-- 
2.11.0

